### ISSUE TITLE
Update Open Liberty's MP Metric Documentation for Metrics 3.0 For 21.0.0.3

### ISSUE THREAD

**Issue by Channyboy** [OpenLiberty/docs#3277]
For 21.0.0.3.

## Update **File 1**:
https://openliberty.io/docs/20.0.0.11/metrics-list.html  
^ Reference link. This link points to 20.0.0.11

~~1. **REMOVE** the following lines.~~ (to keep)
```
Some metric names changed between MicroProfile Metrics 1.0 and MicroProfile Metrics 2.0. Go to the Differences between MicroProfile Metrics 1.0 and 2.0 if you need to map MicroProfile Metrics 2.0 metric information to MicroProfile Metrics 1.0 metric information.

MicroProfile Metrics 2.3 explicitly enables the Performance Monitoring feature. If you use a version earlier than 2.3 of the MicroProfile Metrics feature, then you must also enable the Performance Monitoring feature.
```

2.  ~~**UPDATE** table header to use `mpMetrics-3.0` and update teh Features Required column to point to `mpMetrics-3.0` at `https://openliberty.io/docs/ref/feature/#mpMetrics-3.0.html` when available.~~ 
As suggested by @chirp1  we will use a new column in place of the existing  `Features required` to something that indicate _when the feature was available_.

## Required Information:
Note that *many* metrics would be listed as available since `mpMetrics-2.0` as these metrics had their name changed from 1.x to 2.x. Or that some metrics had their `metric type` changed. A few metrics would be the same  since their first debut in mpMetrics-1.x. Perhaps for these metrics that have been _changed_  we leave a link to the [1.1 and 2.0 diff page](https://openliberty.io/docs/20.0.0.12/reference/metrics-1-dif.html)

**Metric Type changed from 1.x to 2.0 (Hence "available since mpMetrics-2.0")**
```
Metric name: thread.count 			New Value:Gauge, Previous value: Counter
Metric name: thread.daemon.count 	New Value:Gauge, Previous value: Counter
Metric name: thread.max.count 		New Value:Gauge, Previous value: Counter
```

**Name Changed from 1.x to 2.0 (Hence "available since mpMetrics-2.0")**
```
classloader.loadedClasses.count
classloader.loadedClasses.total
classloader.unloadedClasses.total
connectionpool.connectionHandles{datasource=<datasourceName>}
connectionpool.create.total{datasource=<datasourceName>}
connectionpool.destroy.total{datasource=<datasourceName>}
connectionpool.freeConnections{datasource=<datasourceName>}
connectionPool.inUseTime.total{datasource=<datasourceName>}
connectionpool.managedConnections{datasource=<datasourceName>}
connectionpool.queuedRequests.total{datasource=<datasourceName>}
connectionPool.usedConnections.total{datasource=<datasourceName>}
connectionpool.waitTime.total{datasource=<datasourceName>}
gc.time{name=<gcName>}
gc.total{name=<gcName>}
jaxws.client.checkedApplicationFaults.total{endpoint=<endpointName>}
jaxws.client.invocations.total{endpoint=<endpointName>}
jaxws.client.logicalRuntimeFaults.total{endpoint=<endpointName>}
jaxws.client.responseTime.total{endpoint=<endpointName>}
jaxws.client.runtimeFaults.total{endpoint=<endpointName>}
jaxws.client.uncheckedApplicationFaults.total{endpoint=<endpointName>}
jaxws.server.checkedApplicationFaults.total{endpoint=<endpointName>}
jaxws.server.invocations.total{endpoint=<endpointName>}
jaxws.server.logicalRuntimeFaults.total{endpoint=<endpointName>}
jaxws.server.responseTime.total{endpoint=<endpointName>}
jaxws.server.runtimeFaults.total{endpoint=<endpointName>}
jaxws.server.uncheckedApplicationFaults.total{endpoint=<endpointName>}
servlet.request.total{servlet=<servletName>}
servlet.responseTime.total{servlet=<servletName>}
session.activeSessions{appname=<appName>}
session.create.total{appname=<appName>}
session.invalidated.total{appname=<appName>}
session.invalidatedbyTimeout.total{appname=<appName>}
session.liveSessions{appname=<appName>}
threadpool.activeThreads{pool=<poolName>}
threadpool.size{pool=<poolName>}
```

**New Metric in mpMetrics-2.0**
```
cpu.processCpuTime
```

**New metrics in mpMetrics-2.3**
```
REST.request
grpc.client.receivedMessages.total{grpc=<method_signature>}
grpc.client.responseTime.total{grpc=<method_signature>}
grpc.client.rpcCompleted.total{grpc=<method_signature>}
grpc.client.rpcStarted.total{grpc=<method_signature>}
grpc.client.sentMessages.total{grpc=<method_signature>}
grpc.server.receivedMessages.total{grpc=<service_name>}
grpc.server.responseTime.total{grpc=<service_name>}
grpc.server.rpcCompleted.total{grpc=<service_name>}
grpc.server.rpcStarted.total{grpc=<service_name>}
grpc.server.sentMessages.total{grpc=<service_name>}
```
**^^The `grpc.client.*` metrics were introduced with `grpc-1.0` and the `grpc.server.*` ones were introduced with `grpcClient-1.0`**

**New in mpMetrics-3.0**
```
REST.request.unmappedException.total
```

*#The Remaining metrics:*#
```
- Fault Tolerance Metrics were introduced in mpFaultTolerance-1.1 alongside mpMetrics-1.1

- EVERYTHING ELSE has been available since mpMetrics-1.0
```
----------------------------------------------------------

3. **ADD** the new runtime provided REST stat metrics `REST.request.unmappedException.total` **This metric is new as part of the MP Metrics 3.0 specification.**
- See https://github.com/eclipse/microprofile-metrics/issues/533

| MPMETRICS-3.X NAME | MPMETRICS-3.X PROMETHEUS NAME(S) | TYPE| DESCRIPTION | FEATURE(S) REQUIRED |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| REST.request.unmappedException.total |  base_REST_request_unmappedException_total{class="<fully_qualified_class_name>",method="<method_signature>"}   | Counter | The total number of unmapped exceptions that occur from this RESTful resouce method since the start of the server. | [mpMetrics-3.0](https://openliberty.io/docs/ref/feature/#mpMetrics-3.0.html) | 

4.  **UPDATE** description for the existing ` REST.request` metric. The description has been modified as part of the MP Metrics 3.0 specification release.
- See https://github.com/eclipse/microprofile-metrics/issues/533

Given the change to the Metrics list page (i.e in regards to the feature required to features availalbe). As proposed by @Charlotte-Holt. We can show the _old_  description and _new_ description as this change is related to the new metric introduced in item 3.

The old metric description (according to the MP Metrics specification):

```
The number of invocations and total response time of the RESTful resource method since the start of the server. This metric is available in MicroProfile Metrics 2.3 and later.
```

The new metric description(according to the MP Metrics specification):

```
The number of invocations and total response time of this RESTful resource method since the start of the server. The metric will not record the elapsed time nor count of a REST request if it resulted in an *unmapped* exception. Also tracks the highest recorded time duration within the previous completed full minute and lowest recorded time duration within the previous completed full minute
```

A proposed combined description can now be:
```
The metric description for `mpMetrics-2.3` is:
The number of invocations and total response time of the RESTful resource method since the start of the server. This metric is available in MicroProfile Metrics 2.3 and later.

As a result of changes to the metric in MP Metrics 3.0 release, the description has been updated for mpMetrics-3.0:
The number of invocations and total response time of this RESTful resource method since the start of the server. The metric will not record the elapsed time nor count of a REST request if it resulted in an *unmapped* exception. Also tracks the highest recorded time duration within the previous completed full minute and lowest recorded time duration within the previous completed full minute
```

## Update **File 2**:
https://openliberty.io/docs/ref/general/#microservice_observability_metrics.html
1. Update Simple Timer metric section
- See https://github.com/eclipse/microprofile-metrics/issues/

FROM:
```
A simple timer metric tracks the elapsed timing duration and invocation counts.

This type of metric is available beginning in MicroProfile Metrics 2.3. The simple timer is a lightweight alternative to the performance-heavy timer metric.

```

TO:
```
A simple timer metric tracks the elapsed timing duration and invocation counts.

This type of metric is available link:/blog/2020/04/09/microprofile-3-3-open-liberty-20004.html#mra[beginning in MicroProfile Metrics 2.3]. The simple timer is a lightweight alternative to the performance-heavy timer metric.

As of link:<waiting on release blog>[MicroProfile Metrics 3.0] the SimpleTimer metric also tracks the largest and smallest recorded duration of the previous complete minute. A complete minute is defined as 00:00:00.000 - 00:00:59.999.
```

## Update **File 3**:
https://openliberty.io/docs/20.0.0.12/reference/feature/mpMetrics-2.3.html
1. UPDATE a line explaining that mpMetrics-2.3 would automatically start monitor-1.0 feature.

BEFORE:
```
The MicroProfile Metrics feature also can provide access to metrics about server components when it is used in combination with the monitor-1.0 feature.
```

AFTER:
```
MicroProfile Metrics 2.3 explicitly enables the [Performance Monitoring](https://openliberty.io/docs/20.0.0.12/reference/feature/monitor-1.0.html) feature. In prior versions of `mpMetrics` the [Performance Monitoring](https://openliberty.io/docs/20.0.0.12/reference/feature/monitor-1.0.html) had to be independently configured in the `server.xml`. By enabling the [Performance Monitoring](https://openliberty.io/docs/20.0.0.12/reference/feature/monitor-1.0.html), this allows MicroProfile metrics to provide access to metrics about server components. 
```

**Comment by chirp1** [comment]
David and I worked out a variety of updates to documentation description above, particularly so that we tell the user not only what the behavior is with the new feature., but what the behavior was before the new feature was added. We did a little redesign of the table. Approving the feature at https://github.com/OpenLiberty/open-liberty/issues/11460

**Comment by Charlotte-Holt** [comment]
@Channyboy I made all the updates for this issue. Please review: 

- https://docs-draft-openlibertyio.mybluemix.net/docs/21.0.0.3/metrics-list.html
- https://docs-draft-openlibertyio.mybluemix.net/docs/21.0.0.3/reference/feature/mpMetrics-2.3.html
- https://docs-draft-openlibertyio.mybluemix.net/docs/21.0.0.3/microservice-observability-metrics.html

**Comment by Channyboy** [comment]
Looks good! LGTM

**Comment by dmuelle** [comment]
# Peer review

just a few minor things

## Microservice observability with metrics
### Intro
- "An application can be written to produce metrics to be used by operations teams when the application is running in production."
--->
"You can write an application to produce metrics that operations teams use when the application is running in production."

- "They can then visualize the metric data in dashboards, such as Grafana, for analysis."
--->
"They can then visualize the metric data for analysis in dashboards such as https://grafana.com/[Grafana]."

- "When an application is running, you can view your metric data from any browser by accessing a /metrics endpoint, for example, `https://localhost:9443/metrics`."
I think we should clarify that 9443 is an example port number- "... for example, `https://localhost:9443/metrics`, where `9443` is the port number for your application."

### Adding metrics to your applications
Think this section needs to mention Mp Metrics feature
"You can use MicroProfile Metircs with Open Liberty by enabling the feature:mpMetrics[display=Microprofile Metrics] feature in your `server.xml` file.

### Metrics and annotations
-  "...the metric increments each time that an annotated element is hit"- this seems vague- what does "hit" mean in this context?
- is this page missing a section for "histogram"? It's the only metric type that is referenced in the Metrics list that isn't explained here. or anywhere else that I can find.

## Metrics Reference list
- "In the table, the Prometheus metric names are listed after each metric."
---> 
In the following table, the Prometheus metric names are listed after each metric."

## MP Metrics feature description
- just wondering if we need to explicitly call out the difference in enabling monitor when it's already documented in the `Features that this feature enables` autogenerated list on each version of the feature page. If we're calling it out because users are likely to miss that detail that's fine but otherwise sets a weird precedent of repeating whats in the autogen.

**Comment by Charlotte-Holt** [comment]
Microservice observability with metrics
Intro
"An application can be written to produce metrics to be used by operations teams when the application is running in production."
--->
"You can write an application to produce metrics that operations teams use when the application is running in production."

"They can then visualize the metric data in dashboards, such as Grafana, for analysis."
--->
"They can then visualize the metric data for analysis in dashboards such as https://grafana.com/[Grafana]."

"When an application is running, you can view your metric data from any browser by accessing a /metrics endpoint, for example, https://localhost:9443/metrics."
I think we should clarify that 9443 is an example port number- "... for example, https://localhost:9443/metrics, where 9443 is the port number for your application."

Adding metrics to your applications
Think this section needs to mention Mp Metrics feature
"You can use MicroProfile Metircs with Open Liberty by enabling the feature:mpMetrics[display=Microprofile Metrics] feature in your server.xml file.

Metrics and annotations
"...the metric increments each time that an annotated element is hit"- this seems vague- what does "hit" mean in this context?
is this page missing a section for "histogram"? It's the only metric type that is referenced in the Metrics list that isn't explained here. or anywhere else that I can find.
Metrics Reference list
"In the table, the Prometheus metric names are listed after each metric."
--->
In the following table, the Prometheus metric names are listed after each metric."
MP Metrics feature description
just wondering if we need to explicitly call out the difference in enabling monitor when it's already documented in the Features that this feature enables autogenerated list on each version of the feature page. If we're calling it out because users are likely to miss that detail that's fine but otherwise sets a weird precedent of repeating whats in the autogen.

**Comment by dmuelle** [comment]
Just noticed this while reviewing the other issue, but putting a comment here to simplify:

" This will usually be once per method call, but may be more than once if the method call is retried."
--->
"This number is usually once per method call, but might be more if the method call is retried."

**Comment by dmuelle** [comment]
This is usually once per method call, but it might be zero if a circuit breaker prevented execution or more than once if the method call is retried.
--->
This number is usually once per method call, but it might be zero if a circuit breaker prevented execution or more than once per method call if the method call is retried.

**Comment by dmuelle** [comment]
This will usually be once per method call, but may be zero times if the circuit breaker prevents execution or more than once if the method is retried
--->
This number is usually  once per method call, but might be zero  if the circuit breaker prevents execution or more than once per method call if the method is retried.

**Comment by Charlotte-Holt** [comment]
@dmuelle Thanks for the review. All feedback was addressed except the item that I note below:

- https://docs-draft-openlibertyio.mybluemix.net/docs/21.0.0.3/metrics-list.html
- https://docs-draft-openlibertyio.mybluemix.net/docs/21.0.0.3/reference/feature/mpMetrics-2.3.html
- https://docs-draft-openlibertyio.mybluemix.net/docs/21.0.0.3/microservice-observability-metrics.html

**Only thing not addressed:**
- is this page missing a section for "histogram"? It's the only metric type that is referenced in the Metrics list that isn't explained here. or anywhere else that I can find. - **This was initially left off because histogram doesn't have an annotation, but I could see how maybe histogram should be there. I think maybe I can open a separate issue for the OL crew to discuss on the next content call? Whether it should/shouldn't be there, and then update as necessary?**

**Wanted to note:**
- just wondering if we need to explicitly call out the difference in enabling monitor when it's already documented in the Features that this feature enables autogenerated list on each version of the feature page. If we're calling it out because users are likely to miss that detail that's fine but otherwise sets a weird precedent of repeating whats in the autogen. - **I think this is a good point that I didn't think about. I'm wondering whether we should leave the description that mentions Performance Monitoring on 2.2 and earlier? That's what I did in these initial updates. I removed the mention of Performance Monitoring for 2.3 and 3.0. We can discuss if this requires further update.**
